#### QUICK START

[Help](https://github.com/alibaba/euler/wiki/%E5%BF%AB%E9%80%9F%E5%BC%80%E5%A7%8B)

***
##### 数据准备
```
# get ppi_data.py
apt-get update && apt-get install -y curl
# 创建/data
mkdir /data
cd /data
curl -k -O https://raw.githubusercontent.com/alibaba/euler/master/examples/ppi_data.py
pip install networkx==1.11 sklearn -i https://pypi.tuna.tsinghua.edu.cn/simple
```

##### 数据获取
```
python ppi_data.py

# Traceback (most recent call last):
#  File "ppi_data.py", line 32, in <module>
#    from euler.tools import json2dat
#  File "/usr/local/lib/python2.7/dist-packages/euler/__init__.py", line 20, in <module>
#    from euler.python.service import *
#  File "/usr/local/lib/python2.7/dist-packages/euler/python/service.py", line 27, in <module>
#    _LIB = ctypes.CDLL(_LIB_PATH)
#  File "/usr/lib/python2.7/ctypes/__init__.py", line 366, in __init__
#    self._handle = _dlopen(self._name, mode)
# OSError: libhdfs.so.0.0.0: cannot open shared object file: No such file or directory

# ERROR: OSError!!!
```
```
# 环境变量缺失,重新export
export JAVA_HOME=/usr/lib/jvm/default-java
export LD_LIBRARY_PATH=$JAVA_HOME/lib/server:$LD_LIBRARY_PATH
export HADOOP_HOME=/usr/local/hadoop-2.9.2
export LIBRARY_PATH=$HADOOP_HOME/lib/native:$LIBRARY_PATH
export LD_LIBRARY_PATH=$HADOOP_HOME/lib/native:$LD_LIBRARY_PATH
export CLASSPATH=$($HADOOP_HOME/bin/hadoop classpath --glob):$CLASSPATH

# download ppi data..
# unzip data..
# Removed 0 nodes that lacked proper annotations due to networkx versioning issues
# Loaded data.. now preprocessing..
# convert data...
```

##### 模型训练
``` 
python -m tf_euler \
  --data_dir ppi \
  --max_id 56944 --feature_idx 1 --feature_dim 50 --label_idx 0 --label_dim 121 \
  --model graphsage_supervised --mode train

# WARNING: Logging before InitGoogleLogging() is written to STDERR
# I0130 02:06:16.717770 12920 graph_builder.cc:59] Load Done: ppi/ppi_data.dat
# I0130 02:06:17.332028 12920 graph_builder.cc:109] Done: build all sampler
# I0130 02:06:17.332062 12920 graph_builder.cc:112] Graph build finish
# WARNING:tensorflow:From /usr/local/lib/python2.7/dist-packages/tf_euler/python/layers.py:77: __init__ (from tensorflow.python.ops.init_ops) is deprecated and will be removed in a future version.
# Instructions for updating:
# Use tf.initializers.variance_scaling instead with distribution=uniform to get equivalent behavior.
# INFO:tensorflow:Create CheckpointSaverHook.
# INFO:tensorflow:Graph was finalized.
# 2019-01-30 02:06:17.842878: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 AVX512F FMA
# INFO:tensorflow:Running local_init_op.
# INFO:tensorflow:Done running local_init_op.
# INFO:tensorflow:Saving checkpoints for 0 into ckpt/model.ckpt.
# INFO:tensorflow:f1 = 0.3712409, loss = 0.6934097, step = 1
# INFO:tensorflow:f1 = 0.38085398, loss = 0.5414407, step = 21 (1.503 sec)
# ...
# INFO:tensorflow:f1 = 0.59215, loss = 0.43907645, step = 2201 (1.363 sec)
# INFO:tensorflow:Saving checkpoints for 2220 into ckpt/model.ckpt.
```

##### 模型评估
```
python -m tf_euler \
  --data_dir ppi --id_file ppi/ppi_test.id \
  --max_id 56944 --feature_idx 1 --feature_dim 50 --label_idx 0 --label_dim 121 \
  --model graphsage_supervised --mode evaluate

# WARNING: Logging before InitGoogleLogging() is written to STDERR
# I0130 02:15:08.577500 12979 graph_builder.cc:59] Load Done: ppi/ppi_data.dat
# I0130 02:15:09.187830 12979 graph_builder.cc:109] Done: build all sampler
# I0130 02:15:09.187867 12979 graph_builder.cc:112] Graph build finish
# WARNING:tensorflow:From /usr/local/lib/python2.7/dist-packages/tf_euler/python/layers.py:77: __init__ (from tensorflow.python.ops.init_ops) is deprecated and will be removed in a future version.
# Instructions for updating:
# Use tf.initializers.variance_scaling instead with distribution=uniform to get equivalent behavior.
# INFO:tensorflow:Graph was finalized.
# 2019-01-30 02:15:09.476646: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 AVX512F FMA
# INFO:tensorflow:Restoring parameters from ckpt/model.ckpt-2220
# INFO:tensorflow:Running local_init_op.
# INFO:tensorflow:Done running local_init_op.
# f1: 0.619442224503
```

##### embedding输出
```
python -m tf_euler \
  --data_dir ppi \
  --max_id 56944 --feature_idx 1 --feature_dim 50 --label_idx 0 --label_dim 121 \
  --model graphsage_supervised --mode save_embedding

# WARNING: Logging before InitGoogleLogging() is written to STDERR
# I0130 02:45:16.625327 13060 graph_builder.cc:59] Load Done: ppi/ppi_data.dat
# I0130 02:45:17.247097 13060 graph_builder.cc:109] Done: build all sampler
# I0130 02:45:17.247134 13060 graph_builder.cc:112] Graph build finish
# WARNING:tensorflow:From /usr/local/lib/python2.7/dist-packages/tf_euler/python/layers.py:77: __init__ (from tensorflow.python.ops.init_ops) is deprecated and will be removed in a future version.
# Instructions for updating:
# Use tf.initializers.variance_scaling instead with distribution=uniform to get equivalent behavior.
# INFO:tensorflow:Graph was finalized.
# 2019-01-30 02:45:17.536198: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 AVX512F FMA
# INFO:tensorflow:Restoring parameters from ckpt/model.ckpt-2220
# INFO:tensorflow:Running local_init_op.
# INFO:tensorflow:Done running local_init_op.

ll ckpt/

# total 61316
# drwxr-xr-x 2 root root     4096 Jan 30 02:45 ./
# drwxr-xr-x 4 root root     4096 Jan 30 02:06 ../
# -rw-r--r-- 1 root root      130 Jan 30 02:08 checkpoint
# -rw-r--r-- 1 root root 58311808 Jan 30 02:45 embedding.npy
# -rw-r--r-- 1 root root   574951 Jan 30 02:08 events.out.tfevents.1548813977.27d27e6ebbf4
# -rw-r--r-- 1 root root    99006 Jan 30 02:15 events.out.tfevents.1548814509.27d27e6ebbf4
# -rw-r--r-- 1 root root    97918 Jan 30 02:45 events.out.tfevents.1548816317.27d27e6ebbf4
# -rw-r--r-- 1 root root   357263 Jan 30 02:06 graph.pbtxt
# -rw-r--r-- 1 root root   330559 Jan 30 02:45 id.txt
# -rw-r--r-- 1 root root  1313212 Jan 30 02:06 model.ckpt-0.data-00000-of-00001
# -rw-r--r-- 1 root root      897 Jan 30 02:06 model.ckpt-0.index
# -rw-r--r-- 1 root root   173696 Jan 30 02:06 model.ckpt-0.meta
# -rw-r--r-- 1 root root  1313212 Jan 30 02:08 model.ckpt-2220.data-00000-of-00001
# -rw-r--r-- 1 root root      897 Jan 30 02:08 model.ckpt-2220.index
# -rw-r--r-- 1 root root   173696 Jan 30 02:08 model.ckpt-2220.meta
```

##### faiss检索
```
pip install faiss -i https://pypi.tuna.tsinghua.edu.cn/simple

# Collecting faiss
#   Could not find a version that satisfies the requirement faiss (from versions: )
# No matching distribution found for faiss
```
```
# TODO:
import faiss
import numpy as np
embedding = np.load('ckpt/embedding.npy')
index = faiss.IndexFlatIP(128)
index.add(embedding)
print(index.search(embedding[:5], 4))
```
***
END